kv/kvserver: TestReplicaRangefeedErrors failed

kv/kvserver.TestReplicaRangefeedErrors [failed](https://mesolite.cluster.engflow.com/invocations/default/1700532f-5fc4-4aec-a274-6b1939b8a285?testReportRun=1&testReportShard=38&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZzZXJ2ZXI6a3ZzZXJ2ZXJfdGVzdA==) on release-25.4 @ [5bb17b5bdd8bc4cb3a0570c6b3c439e04fc6c961](https://github.com/cockroachdb/cockroach/commits/5bb17b5bdd8bc4cb3a0570c6b3c439e04fc6c961):

Fatal error:

```
panic: runtime error: invalid memory address or nil pointer dereference
	(1) attached stack trace
	  -- stack trace:
	  | runtime.gopanic
	  | 	GOROOT/src/runtime/panic.go:791
	  | github.com/cockroachdb/cockroach/pkg/util/ctxgroup.Group.Wait
	  | 	pkg/util/ctxgroup/ctxgroup.go:151
	  | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.muxRangeFeed
	  | 	pkg/kv/kvclient/kvcoord/dist_sender_mux_rangefeed.go:93
	  | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).RangeFeed
	  | 	pkg/kv/kvclient/kvcoord/dist_sender_rangefeed.go:207
	  | github.com/cockroachdb/cockroach/pkg/kv/kvclient/rangefeed.(*dbAdapter).RangeFeed
	  | 	pkg/kv/kvclient/rangefeed/db_adapter.go:82
	  | github.com/cockroachdb/cockroach/pkg/kv/kvclient/rangefeed.(*RangeFeed).run.func1
	  | 	pkg/kv/kvclient/rangefeed/rangefeed.go:368
	  | [...repeated from below...]
	Wraps: (2) attached stack trace
	  -- stack trace:
	  | runtime.gopanic
	  | 	GOROOT/src/runtime/panic.go:791
	  | runtime.panicmem
	  | 	GOROOT/src/runtime/panic.go:262
	  | runtime.sigpanic
	  | 	GOROOT/src/runtime/signal_unix.go:917
	  | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*activeMuxRangeFeed).start
	  | 	pkg/kv/kvclient/kvcoord/dist_sender_mux_rangefeed.go:284
	  | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*rangefeedMuxer).startSingleRangeFeed
	  | 	pkg/kv/kvclient/kvcoord/dist_sender_mux_rangefeed.go:224
	  | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.divideSpanOnRangeBoundaries
	  | 	pkg/kv/kvclient/kvcoord/dist_sender_rangefeed.go:454
	  | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.divideAllSpansOnRangeBoundaries
	  | 	pkg/kv/kvclient/kvcoord/dist_sender_rangefeed.go:234
	  | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.muxRangeFeed.func2
	  | 	pkg/kv/kvclient/kvcoord/dist_sender_mux_rangefeed.go:90
	  | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.muxRangeFeed.Group.GoCtx.func3
	  | 	pkg/util/ctxgroup/ctxgroup.go:200
	  | golang.org/x/sync/errgroup.(*Group).Go.func1
	  | 	external/org_golang_x_sync/errgroup/errgroup.go:93
	  | runtime.goexit
	  | 	src/runtime/asm_amd64.s:1700
	Wraps: (3) runtime error: invalid memory address or nil pointer dereference
	Error types: (1) *withstack.withStack (2) *withstack.withStack (3) runtime.errorString

```
Stack: 

```
goroutine 80520 [running]:
github.com/cockroachdb/cockroach/pkg/util/ctxgroup.Group.Wait({0xc01a36b580?, {0xda96f00?, 0xc0143eb720?}, 0xc00923f640?})
	pkg/util/ctxgroup/ctxgroup.go:151 +0x134
github.com/cockroachdb/cockroach/pkg/util/ctxgroup.GoAndWait({0xda96f38, 0xc001e8e9c0}, {0xc0044a3b70, 0x2, 0xc001e8e9c0?})
	pkg/util/ctxgroup/ctxgroup.go:224 +0x1bc
github.com/cockroachdb/cockroach/pkg/kv/kvclient/rangefeed.(*RangeFeed).run(0xc00632d6c0, {0xda96f38, 0xc001e8e9c0}, {0xdaf9560, 0xc0073021e0}, 0x0)
	pkg/kv/kvclient/rangefeed/rangefeed.go:394 +0x1431
github.com/cockroachdb/cockroach/pkg/kv/kvclient/rangefeed.(*RangeFeed).start.func1({0xda96f00, 0xc007302230})
	pkg/kv/kvclient/rangefeed/rangefeed.go:254 +0x73c
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func1({0xda96f00, 0xc007302230})
	pkg/util/stop/stopper.go:440 +0x116
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx in goroutine 80517
	pkg/util/stop/stopper.go:438 +0x24f
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestReplicaRangefeedErrors
    test_log_scope.go:171: test logs captured to: outputs.zip/logTestReplicaRangefeedErrors2215917484
    test_log_scope.go:82: use -show-logs to present logs inline
=== RUN   TestReplicaRangefeedErrors/REASON_LOGICAL_OPS_MISSING
    test_server_shim.go:561: DRPC is enabled (override by TestingGlobalDRPCOption)
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=1</code>
 - <code>shard=38</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv-triage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestReplicaRangefeedErrors.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-57056

--- a/pkg/kv/kvclient/kvcoord/dist_sender_mux_rangefeed.go
+++ b/pkg/kv/kvclient/kvcoord/dist_sender_mux_rangefeed.go
@@ -250,8 +250,6 @@ func (m *rangefeedMuxer) startSingleRangeFeed(
 // Upon successfully establishing RPC stream, the ownership of the activeMuxRangeFeed
 // gets transferred to the node event loop goroutine (receiveEventsFromNode).
 func (s *activeMuxRangeFeed) start(ctx context.Context, m *rangefeedMuxer) error {
-	streamID := atomic.AddInt64(&m.seqID, 1)
-
 	// Before starting single rangefeed, acquire catchup scan quota.
 	if err := s.acquireCatchupScanQuota(ctx, m.catchupSem, m.metrics); err != nil {
 		return err
@@ -283,6 +281,12 @@ func (s *activeMuxRangeFeed) start(ctx context.Context, m *rangefeedMuxer) error
 		}
 
 		for !s.transport.IsExhausted() {
+			// Generate a new streamID for each attempt. This allows us to use
+			// LoadAndDelete as a coordination mechanism: whoever successfully deletes
+			// the streamID from the streams map is responsible for handling its
+			// restart.
+			streamID := atomic.AddInt64(&m.seqID, 1)
+
 			args := makeRangeFeedRequest(
 				s.Span, s.token.Desc().RangeID, m.cfg.overSystemTable, s.startAfter, m.cfg.withDiff, m.cfg.withFiltering, m.cfg.withMatchingOriginIDs, m.cfg.consumerID, m.cfg.bulkDelivery)
 			args.Replica = s.transport.NextReplica()
@@ -306,11 +310,17 @@ func (s *activeMuxRangeFeed) start(ctx context.Context, m *rangefeedMuxer) error
 			conn, err := m.establishMuxConnection(ctx, rpcClient, args.Replica.NodeID)
 			s.onConnect(rpcClient, m.metrics)
 
+			var takenOver bool
 			if err == nil {
-				err = conn.startRangeFeed(streamID, s, &args, m.cfg.knobs.beforeSendRequest)
+				err, takenOver = conn.startRangeFeed(streamID, s, &args, m.cfg.knobs.beforeSendRequest)
 			}
 
 			if err != nil {
+				if takenOver {
+					// Another goroutine (restartActiveRangeFeed) took over responsibility
+					// for this activeMuxRangeFeed. Abort without error to let them handle it.
+					return nil
+				}
 				log.VErrEventf(ctx, 1,
 					"RPC error establishing mux rangefeed to r%d, replica %s: %s", args.RangeID, args.Replica, err)
 				if grpcutil.IsAuthError(err) {
@@ -506,13 +516,18 @@ func (m *rangefeedMuxer) receiveEventsFromNode(
 			if active.catchupRes != nil {
 				m.metrics.Errors.RangefeedErrorCatchup.Inc(1)
 			}
-			ms.streams.Delete(event.StreamID)
-			// Restart rangefeed on another goroutine. Restart might be a bit
-			// expensive, particularly if we have to resolve span.  We do not want
-			// to block receiveEventsFromNode for too long.
-			m.g.GoCtx(func(ctx context.Context) error {
-				return m.restartActiveRangeFeed(ctx, active, t.Error.GoError())
-			})
+			// Use LoadAndDelete to coordinate with startRangeFeed. Only spawn
+			// restartActiveRangeFeed if we successfully deleted the streamID.
+			// If startRangeFeed's defer already deleted it, that goroutine is
+			// handling the retry.
+			if _, deleted := ms.streams.LoadAndDelete(event.StreamID); deleted {
+				// Restart rangefeed on another goroutine. Restart might be a bit
+				// expensive, particularly if we have to resolve span. We do not want
+				// to block receiveEventsFromNode for too long.
+				m.g.GoCtx(func(ctx context.Context) error {
+					return m.restartActiveRangeFeed(ctx, active, t.Error.GoError())
+				})
+			}
 			continue
 		}
 
@@ -574,6 +589,9 @@ func (m *rangefeedMuxer) restartActiveRangeFeed(
 
 	if errInfo.evict {
 		active.resetRouting(ctx, rangecache.EvictionToken{})
+		if m.cfg.knobs.afterRoutingReset != nil {
+			m.cfg.knobs.afterRoutingReset()
+		}
 	}
 
 	if errInfo.resolveSpan {
@@ -592,11 +610,16 @@ func (m *rangefeedMuxer) restartActiveRangeFeed(
 }
 
 // startRangeFeed initiates rangefeed for the specified request running
-// on this node connection.  If no error returned, registers stream
-// with this connection.  Otherwise, stream is not registered.
+// on this node connection. If no error returned, registers stream
+// with this connection. Otherwise, stream is not registered.
+//
+// The second return value (takenOver) indicates whether another goroutine
+// (receiveEventsFromNode via restartActiveRangeFeed) has taken over
+// responsibility for this activeMuxRangeFeed. When takenOver is true, the
+// caller should abort without retrying, as the other goroutine is handling it.
 func (c *muxStream) startRangeFeed(
 	streamID int64, stream *activeMuxRangeFeed, req *kvpb.RangeFeedRequest, beforeSend func(),
-) (retErr error) {
+) (retErr error, takenOver bool) {
 	// NB: lock must be held for the duration of this method.
 	// The reasons for this are twofold:
 	//  1. Send calls must be protected against concurrent calls.
@@ -619,20 +642,24 @@ func (c *muxStream) startRangeFeed(
 
 	defer func() {
 		if retErr != nil {
-			// undo stream registration.
-			c.streams.Delete(streamID)
+			// Use LoadAndDelete to detect if someone else already took over. If
+			// receiveEventsFromNode already deleted this streamID and spawned
+			// restartActiveRangeFeed, we should not retry here.
+			if _, found := c.streams.LoadAndDelete(streamID); !found {
+				takenOver = true
+			}
 		}
 	}()
 
 	if c.mu.closed {
-		return net.ErrClosed
+		return net.ErrClosed, false
 	}
 
 	if beforeSend != nil {
 		beforeSend()
 	}
 
-	return c.mu.sender.Send(req)
+	return c.mu.sender.Send(req), false
 }
 
 func (c *muxStream) lookupStream(streamID int64) *activeMuxRangeFeed {
@@ -640,7 +667,7 @@ func (c *muxStream) lookupStream(streamID int64) *activeMuxRangeFeed {
 	return v
 }
 
-// close closes mux stream returning the list of active range feeds.
+// close closes mux stream returning the list of active range feeds to restart.
 func (c *muxStream) close() (toRestart []*activeMuxRangeFeed) {
 	// NB: lock must be held for the duration of this method to synchronize with startRangeFeed.
 	c.mu.Lock()
@@ -648,8 +675,13 @@ func (c *muxStream) close() (toRestart []*activeMuxRangeFeed) {
 
 	c.mu.closed = true
 
-	c.streams.Range(func(_ int64, v *activeMuxRangeFeed) bool {
+	// Collect streams and delete them from the map. Deleting ensures that any
+	// concurrent restartActiveRangeFeed goroutines (spawned by
+	// receiveEventsFromNode before it returned) will see the stream was already
+	// handled via their LoadAndDelete check.
+	c.streams.Range(func(streamID int64, v *activeMuxRangeFeed) bool {
 		toRestart = append(toRestart, v)
+		c.streams.Delete(streamID)
 		return true
 	})
 

--- a/pkg/kv/kvclient/kvcoord/dist_sender_rangefeed.go
+++ b/pkg/kv/kvclient/kvcoord/dist_sender_rangefeed.go
@@ -84,6 +84,9 @@ type rangeFeedConfig struct {
 		captureMuxRangeFeedRequestSender func(nodeID roachpb.NodeID, sender func(req *kvpb.RangeFeedRequest) error)
 		// beforeSendRequest is a mux rangefeed callback invoked prior to sending rangefeed request.
 		beforeSendRequest func()
+		// afterRoutingReset is a mux rangefeed callback invoked after resetRouting
+		// completes. Used for testing race conditions.
+		afterRoutingReset func()
 	}
 }
 
@@ -701,6 +704,14 @@ func TestingWithMuxRangeFeedRequestSenderCapture(
 func TestingWithBeforeSendRequest(fn func()) RangeFeedOption {
 	return optionFunc(func(c *rangeFeedConfig) {
 		c.knobs.beforeSendRequest = fn
+	})
+}
+
+// TestingWithAfterRoutingReset returns a test only option that invokes
+// function after resetRouting completes.
+func TestingWithAfterRoutingReset(fn func()) RangeFeedOption {
+	return optionFunc(func(c *rangeFeedConfig) {
+		c.knobs.afterRoutingReset = fn
 	})
 }
 

--- a/pkg/kv/kvclient/kvcoord/dist_sender_rangefeed_mock_test.go
+++ b/pkg/kv/kvclient/kvcoord/dist_sender_rangefeed_mock_test.go
@@ -8,6 +8,8 @@ package kvcoord
 import (
 	"context"
 	"fmt"
+	"net"
+	"sync/atomic"
 	"testing"
 
 	"github.com/cockroachdb/cockroach/pkg/gossip"
@@ -148,4 +150,165 @@ func TestDistSenderRangeFeedRetryOnTransportErrors(t *testing.T) {
 				require.Error(t, err)
 			})
 	}
+}
+
+// TestMuxRangeFeedTransportRace tests for a data race on activeMuxRangeFeed.transport.
+//
+// The race occurs between:
+//
+//   - Goroutine A: activeMuxRangeFeed.start() retry loop accessing s.transport
+//   - Goroutine B: restartActiveRangeFeed() calling resetRouting() which nils s.transport
+//
+// The race happens because startRangeFeed stores the activeMuxRangeFeed in the
+// streams map before calling Send(). So while Send() is blocked, Recv() can
+// return a RangeFeedError for that streamID. This triggers restartActiveRangeFeed
+// which calls resetRouting(), niling transport. Then Send() unblocks and returns
+// an error, causing start() to continue its retry loop and access the now-nil
+// transport.
+//
+// This test hits a nil pointer exception in the presence of the bug.
+func TestMuxRangeFeedTransportRace(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	clock := hlc.NewClockForTesting(nil)
+	ctx := context.Background()
+
+	stopper := stop.NewStopper()
+	defer stopper.Stop(ctx)
+
+	var cancel func()
+	ctx, cancel = stopper.WithCancelOnQuiesce(ctx)
+	defer cancel()
+
+	rpcContext := rpc.NewInsecureTestingContext(ctx, clock, stopper)
+	g := makeGossip(t, stopper, rpcContext)
+
+	desc := roachpb.RangeDescriptor{
+		RangeID:    1,
+		Generation: 1,
+		StartKey:   roachpb.RKeyMin,
+		EndKey:     roachpb.RKeyMax,
+		InternalReplicas: []roachpb.ReplicaDescriptor{
+			{NodeID: 1, StoreID: 1, ReplicaID: 1},
+			{NodeID: 2, StoreID: 2, ReplicaID: 2},
+		},
+	}
+	for _, repl := range desc.InternalReplicas {
+		require.NoError(t, g.AddInfoProto(
+			gossip.MakeNodeIDKey(repl.NodeID),
+			newNodeDesc(repl.NodeID),
+			gossip.NodeDescriptorTTL,
+		))
+	}
+
+	ctrl := gomock.NewController(t)
+	rangeDB := rangecachemock.NewMockRangeDescriptorDB(ctrl)
+
+	// Synchronization channels.
+	var (
+		// capturedStreamID stores the streamID from Send() so Recv() can use it.
+		capturedStreamID atomic.Int64
+		streamStarted    atomic.Bool
+
+		// sendReady signals that Send() has captured the streamID and is waiting.
+		sendReady = make(chan struct{})
+		// resetDone signals that resetRouting has completed (transport is nil).
+		resetDone = make(chan struct{})
+		// proceedAfterReset allows the afterRoutingReset hook to continue.
+		proceedAfterReset = make(chan struct{})
+	)
+
+	// Create mock client and stream.
+	client := kvpbmock.NewMockRPCInternalClient(ctrl)
+	stream := kvpbmock.NewMockRPCInternal_MuxRangeFeedClient(ctrl)
+
+	client.EXPECT().MuxRangeFeed(gomock.Any()).Return(stream, nil).AnyTimes()
+
+	stream.EXPECT().Send(gomock.Any()).DoAndReturn(func(req *kvpb.RangeFeedRequest) error {
+		capturedStreamID.Store(req.StreamID)
+		close(sendReady)
+		<-resetDone
+		return net.ErrClosed
+	}).AnyTimes()
+
+	stream.EXPECT().Recv().DoAndReturn(func() (*kvpb.MuxRangeFeedEvent, error) {
+		select {
+		case <-sendReady:
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		}
+
+		if streamStarted.Swap(true) {
+			// We only need to return an error the first time. Just block on
+			// subsequent calls.
+			<-ctx.Done()
+			return nil, ctx.Err()
+		}
+
+		streamID := capturedStreamID.Load()
+		return &kvpb.MuxRangeFeedEvent{
+			StreamID: streamID,
+			RangeID:  desc.RangeID,
+			RangeFeedEvent: kvpb.RangeFeedEvent{
+				Error: &kvpb.RangeFeedError{
+					Error: *kvpb.NewError(&kvpb.RangeNotFoundError{RangeID: desc.RangeID}),
+				},
+			},
+		}, nil
+	}).AnyTimes()
+
+	// Transport factory returns mock transports.
+	transportFactory := func() Transport {
+		transport := NewMockTransport(ctrl)
+		transport.EXPECT().IsExhausted().Return(false).AnyTimes()
+		transport.EXPECT().NextReplica().Return(desc.InternalReplicas[0]).AnyTimes()
+		transport.EXPECT().NextInternalClient(gomock.Any()).Return(client, nil).AnyTimes()
+		transport.EXPECT().Release().AnyTimes()
+		return transport
+	}
+
+	rangeDB.EXPECT().RangeLookup(gomock.Any(), roachpb.RKeyMin, kvpb.INCONSISTENT, false).
+		Return([]roachpb.RangeDescriptor{desc}, nil, nil).AnyTimes()
+
+	cachedLease := roachpb.Lease{
+		Replica:  desc.InternalReplicas[0],
+		Sequence: 1,
+	}
+	ds := NewDistSender(DistSenderConfig{
+		AmbientCtx:      log.MakeTestingAmbientCtxWithNewTracer(),
+		Clock:           clock,
+		NodeDescs:       g,
+		RPCRetryOptions: &retry.Options{MaxRetries: 10},
+		Stopper:         stopper,
+		TransportFactory: func(SendOptions, ReplicaSlice) Transport {
+			return transportFactory()
+		},
+		RangeDescriptorDB: rangeDB,
+		Settings:          cluster.MakeTestingClusterSettings(),
+	})
+	ds.rangeCache.Insert(ctx, roachpb.RangeInfo{
+		Desc:  desc,
+		Lease: cachedLease,
+	})
+
+	afterResetFn := func() {
+		close(resetDone)
+		select {
+		case <-proceedAfterReset:
+		case <-ctx.Done():
+		}
+	}
+
+	require.NoError(t, stopper.RunAsyncTask(ctx, "range-feed", func(ctx context.Context) {
+		_ = ds.RangeFeed(
+			ctx,
+			[]SpanTimePair{{Span: roachpb.Span{Key: keys.MinKey, EndKey: keys.MaxKey}}},
+			nil,
+			TestingWithAfterRoutingReset(afterResetFn),
+		)
+	}))
+
+	<-resetDone
+	close(proceedAfterReset)
 }

